build-libs/redirects.js

```diff
@@ -167,20 +168,72 @@ async function getRedirectsForProduct(
 
 	let repo = product
 
-	const rawRedirects = isDeployPreview(product)
-		? fs.readFileSync(path.join(process.cwd(), '../redirects.js'), 'utf-8')
-		: isDeployPreview()
-		? '[]'
-		: await fetchGithubFile({
-				owner: 'hashicorp',
-				repo,
-				path: redirectsPath,
-				ref: latestRef ?? ref,
-		  })
+	/** @type {string} A raw redirects file string to evaluate */
+	let rawRedirects
+
+	/**
+	 * Load the raw redirects
+	 */
+	if (isDeployPreview(product)) {
+		// For deploy previews of this product, load redirects locally,
+		// as authors may be modifying redirects as part of their work.
+		rawRedirects = fs.readFileSync(
+			path.join(process.cwd(), '../redirects.js'),
+			'utf-8'
+		)
+	} else if (isDeployPreview()) {
+		// For deploy previews in "other products", return an empty array,
+		// since we'll be removing the content for this product anyways.
+		rawRedirects = '[]'
+	} else {
+		// Otherwise, such as for production deploys,
+		// load redirects from the product repository.
+		rawRedirects = await fetchGithubFile({
+			owner: 'hashicorp',
+			repo,
+			path: redirectsPath,
+			ref: latestRef ?? ref,
+		})
+	}
+
+	/** @type {Redirect[]} */
 	const parsedRedirects = eval(rawRedirects) ?? []
 
+	/**
+	 * Each of the parsedRedirects should be prefixed with `/{productSlug}`,
+	 * as we're applying these redirects to `developer.hashicorp.com`.
+	 *
+	 * We filter out & ignore any redirects not prefixed with the `product` slug.
+	 *
+	 * We also print out a warning. Note that this warning may not be visible
+	 * to authors, unless they view logs for the preview build they're working on.
+	 */
+
+	/** @type {Redirect[]} */
+	const invalidRedirects = []
+	const validRedirects = parsedRedirects.filter((entry) => {
+		// Redirects must be prefixed with the product slug.
+		const isPrefixed = entry.source.startsWith(`/${product}`)
```

Comment:
For now `isPrefixed` seems like the only validity concern we have. I figure if others come up, probably worth splitting out a `filterValidRedirects` function or something 🤔 But for now this felt alright.